File: 1418.py
class Solution:
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        cur = [0] * k
        N = len(cookies)
        
        def helper(i, c):
            if N - i < c:
                return float('inf')
            if i >= N:
                return max(cur)
            res = float('inf')
            for j in range(k):
                c -= cur[j] == 0
                cur[j] += cookies[i]
                
                res = min(res, helper(i+1, c))
                cur[j] -= cookies[i]
                c += cur[j] == 0
                
            
            return res
        
        return helper(0, k)

File: 1418.java
class Solution {
    int ret = Integer.MAX_VALUE;
    public int distributeCookies(int[] cookies, int k) {
        helper(k, 0, cookies, new int[8]);
        return ret;
        
    }
    public void helper(int k, int cur, int[] arr, int[] plan){
        if(cur == arr.length){
            int mx = 0;
            for(int i : arr)
                mx = Math.max(mx, i);
            ret = Math.min(ret, mx);
            return;
        }
        for(int i = 0; i < k; i++){
            plan[cur] += arr[i];
            helper(k, cur + 1, arr, plan);
            plan[cur] -= arr[i];
        }
    }
}